### PR TITLE
fixed bugs that caused btagSF to be evaluated to zero for some jets:

### PR DESCRIPTION
**Pull request by veelken**
- added workaround for the fact that BTagCalibrationReader::eval_auto_bounds() returns zero
  in case jet pT > 30 or abs(eta) > 2.4
- fixed processing of btagSF for charm and bottom jets

### CODE REVIEW

**Comment by arizzi** [comment]
this is not coherent with what stated here
https://twiki.cern.ch/twiki/bin/view/CMS/BTagCalibration
namely:
In June 2016, starting at CMSSW_8_0_12 and CMSSW_8_1_0, the eval_auto_bounds function is added, providing access to all loaded sys. types and automatically increasing the uncertainty by a factor of two, if the pt value is outside the provided range

**Comment by arizzi** [comment]
are you sure it is both pt and eta out of range that gives zero? it should be only eta.

**Comment by veelken** [comment]
Hi Andrea,

I have checked that the eval_auto_bounds function returns zero in case pT < 30 GeV or abs(eta) > 2.4. I will send an email to the BTV Hypernews and ask the author of that code to confirm.

**Comment by veelken** [comment]
Here is the link to the email : https://hypernews.cern.ch/HyperNews/CMS/get/btag/1478.html

In most of the cases where the SF returns zero, the abs(eta) of the jet is indeed above 2.4. The case where the SF returns zero and the jet pT is below 30 GeV seems more rare (but does occur).

**Comment by arizzi** [comment]
is the discriminator outside 0,1 boundary?

Il 30 ott 2017 9:30 AM, "Christian Veelken" <notifications@github.com> ha
scritto:

**Comment by veelken** [comment]
Nope. It may be, however, that only in some bins of the discriminator distribution the SF evaluates to zero for jets of pT < 30 GeV.

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 1]
```diff
@@ -184,37 +192,42 @@ def getSF(self, pt, eta, flavor, syst = 'central', wp = 'M', measurement_type =
...
+
+        for central_or_syst in self.central_and_systs:
             scale_factors = []
-            for central_or_syst in self.central_and_systs:
-                sf = self.getSF(jet.pt, jet.eta, jet.partonFlavour, central_or_syst, 'M', 'auto', False, discr)
-                #print " %s = %1.3f" % (self.branchNames[central_or_syst], sf)
+            for idx, jet in enumerate(jets):
+                sf = self.getSF(jet.pt, jet.eta, jet.partonFlavour, central_or_syst, 'M', 'auto', False, getattr(jet, discr))
+                if self.verbose > 0 and sf < 0.01:
+                    print("jet #%i: pT = %1.1f, eta = %1.1f, discr = %1.3f, flavor = %i" % (idx, jet.pt, jet.eta, getattr(jet, discr), jet.partonFlavour))
```
Comment (arizzi) [inline python/postprocessing/modules/btv/btagSFProducer.py]: here I would add if sf < 0.01 then sf =1

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 2]
```diff
@@ -184,37 +192,42 @@ def getSF(self, pt, eta, flavor, syst = 'central', wp = 'M', measurement_type =
...
+            #print(self.branchNames_central_and_systs[central_or_syst], scale_factors)
+        for central_or_syst in self.central_and_systs_shape_corr:
             scale_factors = []
-            for central_or_syst in self.central_and_systs_shape_corr:
-                sf = self.getSF(jet.pt, jet.eta, jet.partonFlavour, central_or_syst, 'shape_corr', 'auto', True, discr)
-                #print " %s = %1.3f" % (self.branchNames[central_or_syst], sf)
+            for idx, jet in enumerate(jets):
+                sf = self.getSF(jet.pt, jet.eta, jet.partonFlavour, central_or_syst, 'shape_corr', 'auto', True, getattr(jet, discr))
+                if self.verbose > 0 and sf < 0.01:
+                    print("jet #%i: pT = %1.1f, eta = %1.1f, discr = %1.3f, flavor = %i" % (idx, jet.pt, jet.eta, getattr(jet, discr), jet.partonFlavour))
```
Comment (arizzi) [inline python/postprocessing/modules/btv/btagSFProducer.py]: also here

**Comment by veelken** [comment]
Hi Andrea,

I have updated the code following your requests and what we discussed on Skype. Please have a look at the code again and merge PR #23 if everything is ok.

**Pull request merged**